cpp/src/arrow/testing/gtest_util.cc

```diff
@@ -907,7 +910,7 @@ class GatingTask::Impl : public std::enable_shared_from_this<GatingTask::Impl> {
   double timeout_seconds_;
   Status status_;
   bool unlocked_;
-  int num_launched_ = 0;
+  std::atomic<int> num_launched_;
```

Comment:
Nit: since we don't care about memory order, I think we can change this line to
`std::atomic<int> num_launched_{0}`, and leave all other codes untouched.